### ISSUE TITLE
[DEV] automatically scan images using a variety of compliance standards in github CI

### ISSUE THREAD

**Issue by RoyalOughtness** [secureblue/secureblue#497]
Example: https://candrews.integralblue.com/2023/09/scap-security-and-compliance-scanning-of-docker-images-in-github-actions-and-gitlab-ci/

References:

https://public.cyber.mil/stigs/
https://www.open-scap.org/
https://www.cisecurity.org/cis-benchmarks
https://github.com/aquasecurity/trivy

**Comment by r3m8** [comment]
Hello,

Thank you for prioritizing this essential functionality. I'd like to know what strategy SecureBlue intends to take to ensure compliance testing ; as I've been responsible for compliance issues professionally for several years, I could contribute to OpenSCAP if necessary.

I've seen that Trivy has been added to automatically detect CVEs, and that's pretty ideal for a CI. However, would you also like to keep Trivy for compliance operations ? Trivy-action CI doesn't fully support compliance at the moment, and seems to be geared more towards cloud and containerized environments. The CIS benchmark in place, unless I'm mistaken, is based on kubernetes (1.23) and not a true full system.

It seems to have two possibilities, where I think the second is the most interesting :

- Refining Trivy with custom rego rules, which seems possible but undoubtedly extremely time-consuming. CIS benchmark work or any other certification does not seem to be available in either Trivy or OPA.
- Integrate OpenSCAP and RHCOS4's ComplianceAsCode content to perform the compliance audit, leveraging already existing work for all available rules/checks with ANSSI-BP-028, BSI-2022, DISA STIG, ACSC, NIST 800-53 and other.

Profiles are available here: https://github.com/ComplianceAsCode/content/tree/master/products/rhcos4

oscap-chroot or oscap-podman can be used for rootfs: https://static.open-scap.org/openscap-1.4.1/oscap_user_manual.html#_scanning_arbitrary_filesystems_using_oscap_chroot

However, OpenSCAP doesn't support sarif format, so rendering pages are generally available in HTML/XML, which I agree isn't very readable or integrated.

So I tested with oscap-chroot ; many checks are marked as “notapplicable” because they can only be performed in a running system : no kernel available, no systemd initialization, no d-bus socket, need to install additional packages, no access to physical hardware.

```
I: oscap: Evaluating definition 'oval:ssg-system_with_kernel:def:1': (null). [oscap(3873):oscap(7f16d143ab80):oval_resultDefinition.c:159:oval_result_definition_eval]
I: oscap: Definition 'oval:ssg-system_with_kernel:def:1' evaluated as false. [oscap(3873):oscap(7f16d143ab80):oval_resultDefinition.c:170:oval_result_definition_eval]
I: oscap: Rule 'xccdf_org.ssgproject.content_rule_audit_rules_time_watch_localtime' is not applicable. [oscap(3873):oscap(7f16d143ab80):xccdf_policy.c:1092:_xccdf_policy_rule_evaluate]
Result  notapplicable
```

I'm not sure if the Github Action CI allows you to run KVM virtual machines, but I can give it a try.

Have a nice day

**Comment by r3m8** [comment]
**Secureblue now runs on Github Actions as a virtual machine**. This means that all compliance operations can be carried out without error.

The repository: https://github.com/r3m8/github-actions-secureblue
The workflow: https://github.com/r3m8/github-actions-secureblue/actions/runs/15751658528/job/44397878417

In detail:

- A [free public Github instance](https://docs.github.com/fr/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories); 4 vCPUs, 16GB RAM, 14GB SSD, x64
- The latest version of Ubuntu 24.04 LTS
- Installation of qemu & libvirt for easy deployment
- KVM configuration & full AppArmor support
- Ingition file generation with Butane 0.24.0
- Rebase to secureblue with [modified installation script](https://github.com/r3m8/github-actions-secureblue/blob/main/install.sh)

```
Run ssh -o ConnectTimeout=200 -o StrictHostKeyChecking=no -i ~/.ssh/id_ed25519 core@${VM_IP} "cat /etc/os-release"
NAME="secureblue"
VERSION="42.20250609.2.0 (CoreOS)"
RELEASE_TYPE=stable
ID="fedora"
ID_LIKE="secureblue"
VERSION_ID=42
VERSION_CODENAME=""
PLATFORM_ID="platform:f42"
PRETTY_NAME="secureblue (powered by Fedora Atomic)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:secureblue:secureblue:42"
HOME_URL="https://github.com/secureblue/secureblue"
DOCUMENTATION_URL="https://github.com/secureblue/secureblue/tree/live/docs"
SUPPORT_URL="https://github.com/secureblue/secureblue/issues"
BUG_REPORT_URL="https://github.com/secureblue/secureblue/issues"
SUPPORT_END=2026-05-13
VARIANT="CoreOS"
VARIANT_ID=securecore-zfs-main-hardened
OSTREE_VERSION='42.20250609.2.0'
```

All that remains is to adapt OpenSCAP for secureblue, based on [RHCOS4 current policy](https://github.com/ComplianceAsCode/content/tree/master/products/rhcos4), whether this approach is suitable and accepted by maintainers.

**Comment by RoyalOughtness** [comment]
@r3m8 Thank you for digging into this! Yes, if you have time, running VMs of secureblue in Github Actions is something we desperately need. Not just for compliance but also for integration testing. 

> as I've been responsible for compliance issues professionally for several years, I could contribute to OpenSCAP if necessary.

That would be awesome!

@HastD FYI

**Comment by HastD** [comment]
Very cool! I'll take a look and see if I can adapt this for integration testing of [run0edit](https://github.com/HastD/run0edit), which is difficult to test in a simple container environment.

**Comment by r3m8** [comment]
I was able to make progress on compliance. Here are the first results:

- [ANSSI-BP-028](https://r3m8.github.io/oscap-secureblue/content_profile_anssi_bp28_high-report.html) (high)
- [BSI-2022](https://r3m8.github.io/oscap-secureblue/content_profile_bsi-2022-report.html) (incomplete from upstream)
- [NIST 800-53](https://r3m8.github.io/oscap-secureblue/content_profile_high-rev-4-report.html) (high)
- [DISA STIG](https://r3m8.github.io/oscap-secureblue/content_profile_stig-v2r2-report.html)

There's quite a bit of adaptation to be done for secureblue, but nothing impossible. The Red Hat team doesn't seem to have implemented all the profiles available for RHCOS either; we can use code from other policies to generate profiles specific to secureblue.

I'd be interested in your comments on the policy criteria. Some of them seem inadequate, especially when “sudo” is considered a non-conformity. It's important to understand that these recommendations were originally made for businesses and public organizations, which encourages strengthening security without making major hardening as secureblue would.

**Comment by r3m8** [comment]
I believe that secureblue *will never fully comply* with compliance standards because it calls into question the fundamentals of security (using run0 instead of sudo for example).

What I propose is to create a tailoring file for each compliance standard (ANSSI, PCI-DSS, CIS, BSI...) to enable us to remove certain compliance elements that are not applicable to secureblue, or already mitigated by another approach:

> Tailoring File is a XML file very much like SCAP Content. It represents a customization of compliance defined in a SCAP Content XML file. Additional details and instructions how to create such a file are available in the official openscap [documentation](https://www.open-scap.org/resources/documentation/customizing-scap-security-guide-for-your-use-case/).

This allows us to avoid a new compilation without having to make changes to [CaC](https://github.com/ComplianceAsCode/content) profiles directly. We can use [autotailor](https://www.mankier.com/8/autotailor) to create XML files based on the [reference schema](https://github.com/ComplianceAsCode/schemas/blob/main/tailoring/schema.json). If it's still relevant and worth the cost (I'm not sure):

- Where can we place these tailoring files? In a new or existing repository?
- Would it be possible to write the content of tailoring files in `yaml` and then convert them to `json` with schema validation in CI? `autotailor` doesn't natively support yaml
- For reference, [here is the list of compliance standards](https://github.com/ComplianceAsCode/content/tree/master/controls) supported in CaC

But even if we have to disable a certain number of rules, it's safe to say that secureblue won't be very compliant and can never be used in certain environments, such as PCI-DSS, without significant changes to the criteria of compliance standards.

**Comment by RoyalOughtness** [comment]
@r3m8 yeah, the idea behind this issue was never to comply with the standards, it was to find relevant issues by scanning using the relevant portions of available standards.

**Comment by RoyalOughtness** [comment]
https://www.mankier.com/8/oscap-im
